src/dtree.ml

```diff
@@ -0,0 +1,648 @@
+      (* Create edge *)
+      F.fprintf ppf "@ %d -- %d [label=<%a>];" father_l tag pp_dotterm swon ;
+      ConsMap.iter (fun s e -> write_tree tag (DotCons(s)) e) children ;
+      (match default with None -> () | Some(tr) -> write_tree tag DotDefa tr) ;
+    | Fetch(store, next) ->
+      incr nodecount ;
+      let tag = !nodecount in
+      let store_tag = if store then "s" else "" in
+      F.fprintf ppf "@ %d [label=<%s> shape=\"diamond\"]" tag store_tag ;
+      F.fprintf ppf "@ %d -- %d [label=<%a>];" father_l tag pp_dotterm swon ;
+      write_tree tag DotDefa next
+    | Fail               ->
+      incr nodecount ;
+      F.fprintf ppf "@ %d -- %d [label=\"!\"];" father_l !nodecount
+  in
+  begin
+    match tree with
+    (* First step must be done to avoid drawing a top node. *)
+    | Node({ swap ; children = ch ; store ; default }) ->
+       F.fprintf ppf "@ 0 [label=\"%d\"%s];"
+         swap (if store then " shape=\"box\"" else "") ;
+       ConsMap.iter (fun sw c -> write_tree 0 (DotCons(sw)) c) ch ;
+       (match default with None -> () | Some(tr) -> write_tree 0 DotDefa tr)
+    | Leaf(_)                                -> ()
+    | _                                      -> assert false
+  end ;
+  F.fprintf ppf "@.}@\n@?" ;
+  close_out ochan
+
+(** {3 Clause matrix and pattern matching problem} *)
+
+(** A clause matrix encodes a pattern matching problem.  The clause matrix {i
+    C} can be denote {i C = P → A} where {i P} is a {e pattern matrix} and {i
+    A} is a column of {e actions}.  Each line of a pattern matrix is a pattern
+    to which is attached an action.  When reducing a term, if a line filters
+    the term, or equivalently the term matches the pattern, the term is
+    rewritten to the action. *)
+module ClauseMat =
+struct
+  (** A component of the matrix, a term along with its position in the top
+      term. *)
+  type component = term * Sub.t
+
+  (** [pp_component o c] prints component [c] to channel [o]. *)
+  let pp_component : component pp = fun oc (te, pos) ->
+    Format.fprintf oc "@[<h>%a@ %@@ %a@]" Print.pp te Basics.Subterm.pp pos
+
+  (** A redefinition of the rule type.
+
+      {b Note} that {!type:array} is used while {!module:ReductionStack} could
+      be used because {!val:pick_best_among} goes through all items of a rule
+      anyway ([max(S) = Θ(|S|)]). *)
+  type rule = { lhs : component array
+              (** Left hand side of a rule.   *)
+              ; rhs : action
+              (** Right hand side of a rule. *)
+              ; variables : int SubMap.t
+              (** Mapping from positions of variable subterms in [lhs] to a
+                  slot in a term env. *)}
+
+  (** Type of a matrix of patterns.  Each line is a row having an attached
+      action. *)
+  type t = { clauses : rule list
+           (** The rules. *)
+           ; var_catalogue : Sub.t list
+           (** Contains positions of terms in {!field:lhs} that can be used
```

Comment:
I don't understand what is var_catalogue. The comment refers to a lhs field but the other field is a list of clauses, so multiple lhs's.

Reply:
I hope it's clearer now